### DISCUSSION TITLE
在論文中Multi-vector Queries的詳細解釋

### DISCUSSION THREAD

**Discussion by TwPanTsu** [milvus-io/milvus#23057] [Q&A and General discussion] [answered]
可以詳細又簡單的方式讓我理解甚麼是Vector fusion和Iterative merging嗎 我在論文中看不大懂 謝謝協助
我的理解是vector fusion利用weight sum和inner product把multiple vector合為single vector來找top-k
Iterative merging利用NRA Algo優化來找top-k
可以舉的例子讓我了解詳細過程嗎? 我非常想要介紹你們的論文給其他人!
[milvus.pdf](https://github.com/milvus-io/milvus/files/11086690/milvus.pdf)

**Comment by PwzXxm** [comment, marked as answer]
@TwPanTsu asks about *Vector Fusion* and *Iterative Merging* for Multi-vector Queries in Section 4.2 of the SIGMOD'21 Milvus paper.

Your understanding is basically correct. I will use the same notation as the paper. The paper states the problem is

> Formally, each entity contains 𝜇 vectors v0, v1, ..., v𝜇−1. Then a multi-vector query finds top-𝑘 entities according to an aggregated scoring function 𝑔 over the similarity function 𝑓 (e.g., inner product) of each individual vector v𝑖. Specifically, the similarity of two entities 𝑋 and 𝑌 is computed as 𝑔(𝑓(𝑋.v0,𝑌.v0),...,𝑓(𝑋.v𝜇−1,𝑌.v𝜇−1)) where 𝑋 .v𝑖 means the vector v𝑖 of the entity 𝑋. 

*Vector fusion* concatenates $\mu$ vectors $v_0, v_1, ..., v_{𝜇−1}$ in an arbitrary entity $e$. For example, if $v_0 = [1,2], v_1 = [3,4]$, then we have $v = [1,2,3,4]$. You can imagine that when a query with multi-vectors comes in, it also needs to be processed before the query because it is no longer compatible with the concatenated base vectors. Since the base vectors are already 'fused' and since the inner product is decomposable, we can apply aggregation scoring function $g$ on each multi-vectors first, concatenate them so that they have the same dimension as the base vectors, and then query it. The weighted sum you mentioned is just an example of the aggregated scoring function in the paper.

*Iterative merging* deals with cases the similarity function is not decomposable and data are not normalized. We first initialize $k' = k$. Then it performs vector queries on each multi-vectors and checks if we can deduce $k$ results via the NRA algorithm; if not, as the name suggests, we double the $k'$, perform queries, and check iteratively. Maybe you can check out this paper about the NRA algorithm if necessary: https://dl.acm.org/doi/10.1145/375551.375567

**Reply by TwPanTsu** [reply]
@PwzXxm So, Vector fusion is like using aggregation scoring function g to process every vector on the same scale first. And then concatenate together.
Last, using concatenated vectors to process like single vector query. Am I right ?
Can you give me an example that use Iterative merging instead of Vector fusion ? Because it looks like Vector fusion have more restrictions.

**Reply by TwPanTsu** [reply]
@PwzXxm and aggregation scoring function g in vector fusion is a must or not?

**Reply by PwzXxm** [reply]
> So, Vector fusion is like using aggregation scoring function g to process every vector on the same scale first. And then concatenate together. Last, using concatenated vectors to process like single vector query. Am I right ? 

Yes

> Can you give me an example that use Iterative merging instead of Vector fusion ? Because it looks like Vector fusion have more restrictions.

It's more like the other way around. Vector fusion (VF) has restrictions in which the similarity function is decomposable and the data is normalized. However, Iterative merging (IM) has no such restriction. Scenarios VF can deal with can also be performed under IM, but sadly not the other way around. For example, if the similarity metric is *Euclidean distance* then VF is not applicable since it is not decomposable.

**Reply by PwzXxm** [reply]
> and aggregation scoring function g in vector fusion is a must or not?

When you have $\mu$ outputs where you computed the similarity function on each sub-vectors 𝑓(𝑋.v0,𝑌.v0),...,𝑓(𝑋.v𝜇−1,𝑌.v𝜇−1), you need something to reduce it to a single value so that you can easily compare it with other $\mu$ outputs you obtained comparing other two entities, right? If you have a way to compare $\mu$ outputs, then it might also work depending on your application. Does it make sense?

**Reply by TwPanTsu** [reply]
yes thank you, I think I understand.
